File: apps/api/src/services/worker.service.ts
import { Prisma } from "@prisma/client";
import prisma from "../lib/prisma";
import { AppError } from "../middleware/errorHandler";
import { verifyAadhaar } from "../lib/digilocker";
import { logger } from "../lib/logger";

function haversineDistance(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLng = ((lng2 - lng1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) * Math.cos((lat2 * Math.PI) / 180) *
    Math.sin(dLng / 2) * Math.sin(dLng / 2);
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function normalizeSkill(skill: string): string {
  return skill.trim().toLowerCase().replace(/_/g, "-").replace(/\s+/g, "-");
}

function matchesSkills(workerSkills: string[], requiredSkills?: string[]): boolean {
  if (!requiredSkills || requiredSkills.length === 0) return true;
  const workerSet = new Set(workerSkills.map(normalizeSkill));
  return requiredSkills.some((skill) => workerSet.has(normalizeSkill(skill)));
}

function workerMatchScore(worker: {
  avgRating: number;
  totalJobs: number;
  distanceKm: number;
}, radiusKm: number): number {
  const distanceScore = Math.max(0, 1 - worker.distanceKm / Math.max(radiusKm, 1)) * 60;
  const ratingScore = (Number(worker.avgRating || 0) / 5) * 25;
  const trackRecordScore = Math.min(Number(worker.totalJobs || 0) / 200, 1) * 15;
  return Math.round((distanceScore + ratingScore + trackRecordScore) * 10) / 10;
}

export async function registerWorker(
  userId: string,
  data: {
    skillTags: string[];
    bio?: string;
    experienceYears?: number;
    coopId?: string;
    latitude?: number;
    longitude?: number;
    workAddress?: string;
    kycDocumentUrl?: string;
    aadhaarNumber: string;
    aadhaarName?: string;
    digilockerRef?: string;
  }
): Promise<any> {
  const existing = await prisma.workerProfile.findUnique({ where: { userId } });
  if (existing) throw new AppError("Worker profile already exists", 409);
  let coopId = data.coopId;
  if (data.coopId) {
    const coop = await prisma.coOp.findUnique({ where: { id: data.coopId } });
    if (!coop) throw new AppError("Co-op not found", 404);
  } else if (data.latitude !== undefined && data.longitude !== undefined) {
    // Auto-assign to the nearest active co-op based on the worker's work location.
    const coops = await prisma.coOp.findMany({ where: { isActive: true } });
    let best: { id: string; d: number } | null = null;
    for (const c of coops) {
      const d = haversineDistance(data.latitude, data.longitude, Number(c.latitude), Number(c.longitude));
      if (d <= (c.radiusKm || 10) && (!best || d < best.d)) {
        best = { id: c.id, d };
      }
    }
    coopId = best?.id;
  }
  // DigiLocker verification (dummy — returns verified for any 12-digit Aadhaar)
  const digilockerResult = await verifyAadhaar(data.aadhaarNumber);
  const profile = await prisma.workerProfile.create({
    data: {
      userId,
      skillTags: data.skillTags,
      bio: data.bio,
      experienceYears: data.experienceYears || 0,
      coopId,
      latitude: data.latitude,
      longitude: data.longitude,
      workAddress: data.workAddress,
      kycDocumentUrl: data.kycDocumentUrl,
      kycStatus: data.kycDocumentUrl ? "UNDER_REVIEW" : "PENDING",
      aadhaarNumber: data.aadhaarNumber,
      aadhaarVerified: digilockerResult.verified,
      aadhaarName: digilockerResult.name || data.aadhaarName,
      aadhaarDob: digilockerResult.dob,
      digilockerRef: data.digilockerRef || `DL-${Date.now()}`,
      phoneVerified: true,
      // New workers start pending admin approval — never instantly active.
      status: "PENDING_ADMIN_APPROVAL",
    },
    include: { user: { select: { id: true, name: true, phone: true } }, coop: true },
  });
  await prisma.user.update({ where: { id: userId }, data: { role: "WORKER" } });
  return formatWorker(profile);
}

export async function updateLocation(workerId: string, lat: number, lng: number): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { latitude: lat, longitude: lng },
    include: { user: { select: { id: true, name: true } } },
  });
  return formatWorker(updated);
}

export async function setAvailability(workerId: string, data: { isAvailable: boolean; isOnDuty: boolean }): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  if (data.isOnDuty && profile.status !== "VERIFIED") {
    throw new AppError("Worker must be approved before going on duty", 403);
  }
  if (data.isOnDuty && !data.isAvailable) throw new AppError("Cannot be on duty if not available", 400);
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { isAvailable: data.isAvailable, isOnDuty: data.isOnDuty },
    include: { user: { select: { id: true, name: true } } },
  });
  return formatWorker(updated);
}

export async function getWorkerProfile(workerId: string): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({
    where: { id: workerId },
    include: {
      user: { select: { id: true, name: true, phone: true, avatarUrl: true } },
      coop: { select: { id: true, name: true } },
      reviewsReceived: {
        include: {
          author: { select: { id: true, name: true } },
          booking: { select: { id: true, bookingRef: true } },
        },
        orderBy: { createdAt: "desc" }, take: 10,
      },
    },
  });
  if (!profile) throw new AppError("Worker profile not found", 404);
  return formatWorker(profile);
}

export async function getWorkerEarnings(workerId: string): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  const completedBookings = await prisma.booking.findMany({
    where: { workerId, status: "COMPLETED" },
    select: { id: true, bookingRef: true, finalPrice: true, workerPayout: true, completedAt: true, service: { select: { name: true, categoryName: true } } },
    orderBy: { completedAt: "desc" },
  });
  const transactions = await prisma.walletTransaction.findMany({
    where: { workerId }, orderBy: { createdAt: "desc" }, take: 50,
  });
  const thisMonth = new Date(); thisMonth.setDate(1); thisMonth.setHours(0, 0, 0, 0);
  const monthlyEarnings = completedBookings
    .filter(b => b.completedAt && b.completedAt >= thisMonth)
    .reduce((sum, b) => sum + Number(b.workerPayout || 0), 0);
  return {
    totalEarnings: Number(profile.totalEarnings), walletBalance: Number(profile.walletBalance),
    totalJobs: profile.totalJobs, avgRating: profile.avgRating,
    monthlyEarnings: Math.round(monthlyEarnings * 100) / 100,
    recentBookings: completedBookings.slice(0, 10).map(b => ({ ...b, finalPrice: Number(b.finalPrice || 0), workerPayout: Number(b.workerPayout || 0) })),
    recentTransactions: transactions.slice(0, 20).map(t => ({ ...t, amount: Number(t.amount), balanceAfter: Number(t.balanceAfter) })),
  };
}

export async function verifyWorker(workerId: string, kycData: { aadhaarNumber: string }): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  const result = await verifyAadhaar(kycData.aadhaarNumber);
  if (!result.verified) throw new AppError("Aadhaar verification failed", 400);
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { status: "VERIFIED", kycStatus: "VERIFIED", aadhaarVerified: true, digilockerRef: `DL-${Date.now()}` },
    include: { user: { select: { id: true, name: true } } },
  });
  logger.info(`Worker ${workerId} verified via DigiLocker`);
  return formatWorker(updated);
}

// Co-op admin approval gate (separate from automated DigiLocker KYC check).
export async function approveWorker(workerId: string, coopId: string, note?: string): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  if (profile.coopId && profile.coopId !== coopId) {
    throw new AppError("Worker does not belong to this co-op", 403);
  }
  if (profile.status === "SUSPENDED" || profile.status === "DEACTIVATED") {
    throw new AppError("Cannot approve a suspended/deactivated worker", 400);
  }
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { status: "VERIFIED", kycStatus: "VERIFIED", aadhaarVerified: true },
    include: { user: { select: { id: true, name: true } } },
  });
  logger.info(`Worker ${workerId} approved by co-op ${coopId}${note ? ` (note: ${note})` : ""}`);
  return formatWorker(updated);
}

export async function rejectWorker(workerId: string, coopId: string, reason: string): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  if (profile.coopId && profile.coopId !== coopId) {
    throw new AppError("Worker does not belong to this co-op", 403);
  }
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { status: "SUSPENDED", kycStatus: "REJECTED" },
    include: { user: { select: { id: true, name: true } } },
  });
  logger.info(`Worker ${workerId} rejected by co-op ${coopId}: ${reason}`);
  return formatWorker(updated);
}

export async function searchWorkers(
  lat: number, lng: number, radiusKm: number, skillTags?: string[], coopId?: string
): Promise<any[]> {
  const where: any = { status: "VERIFIED", latitude: { not: null }, longitude: { not: null }, user: { isActive: true } };
  if (coopId) where.coopId = coopId;
  const candidates = await prisma.workerProfile.findMany({
    where, include: { user: { select: { id: true, name: true } }, coop: { select: { name: true } } }, take: 100,
  });
  return candidates
    .map(w => ({
      workerId: w.id, workerName: w.user.name, coopName: w.coop?.name || null,
      skillTags: w.skillTags, avgRating: Number(w.avgRating), totalJobs: w.totalJobs,
      bio: w.bio, experienceYears: w.experienceYears, isAvailable: w.isAvailable, isOnDuty: w.isOnDuty,
      distanceKm: Math.round(haversineDistance(lat, lng, w.latitude!, w.longitude!) * 100) / 100,
    }))
    .filter(w => matchesSkills(w.skillTags, skillTags))
    .filter(w => w.distanceKm <= radiusKm)
    .map((w) => ({
      ...w,
      etaMinutes: Math.max(10, Math.round(w.distanceKm * 5)),
      matchScore: workerMatchScore(w, radiusKm),
    }))
    .sort((a, b) => b.matchScore - a.matchScore || a.distanceKm - b.distanceKm)
    .slice(0, 100);
}

function formatWorker(worker: any): any {
  return { ...worker, totalEarnings: worker.totalEarnings ? Number(worker.totalEarnings) : undefined, walletBalance: worker.walletBalance ? Number(worker.walletBalance) : undefined };
}

export default { registerWorker, updateLocation, setAvailability, getWorkerProfile, getWorkerEarnings, verifyWorker, approveWorker, rejectWorker, searchWorkers };

File: backend/app/services/worker_service.py
import re
import time
import uuid
from datetime import datetime, timezone

from ..db import db, row_to_dict, rows_to_dicts
from ..errors import AppError
from ..utils import haversine_km, matches_skills, normalize_skill, worker_match_score, num, deep_serialize, now_utc



def _now_ts() -> str:
    return f"DL-{int(time.time() * 1000)}"


async def register_worker(user_id: str, data: dict) -> dict:
    existing = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE "userId"=$1', user_id)
    if existing is not None:
        raise AppError("Worker profile already exists", 409)

    coop_id = data.get("coopId")
    if coop_id:
        coop = await db.fetchrow('SELECT * FROM "CoOp" WHERE id=$1', coop_id)
        if coop is None:
            raise AppError("Co-op not found", 404)
    elif data.get("latitude") is not None and data.get("longitude") is not None:
        coops = await db.fetch('SELECT * FROM "CoOp" WHERE "isActive"=true')
        best = None
        for c in coops:
            d = haversine_km(
                data["latitude"], data["longitude"],
                num(c["latitude"]), num(c["longitude"]),
            )
            if d <= num_or(c["radiusKm"], 10) and (best is None or d < best[1]):
                best = (c["id"], d)
        coop_id = best[0] if best else None

    aadhaar_name = data.get("aadhaarName")
    aadhaar_dob = data.get("aadhaarDob")
    aadhaar_verified = bool(re.fullmatch(r"\d{12}", data["aadhaarNumber"]))
    now = now_utc()
    profile_id = str(uuid.uuid4())
    kyc_url = data.get("kycDocumentUrl")
    await db.execute(
        """
        INSERT INTO "WorkerProfile" (
            id, "userId", "skillTags", bio, "experienceYears", "coopId",
            latitude, longitude, "workAddress", "kycDocumentUrl", "kycStatus",
            "aadhaarNumber", "aadhaarVerified", "aadhaarName", "aadhaarDob",
            "digilockerRef", "phoneVerified", status, "createdAt", "updatedAt"
        ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,true,$17,$18,$18)
        """,
        profile_id,
        user_id,
        data.get("skillTags") or [],
        data.get("bio"),
        data.get("experienceYears") or 0,
        coop_id,
        data.get("latitude"),
        data.get("longitude"),
        data.get("workAddress"),
        kyc_url,
        "VERIFIED" if kyc_url else "UNDER_REVIEW",
        data["aadhaarNumber"],
        aadhaar_verified,
        aadhaar_name or data.get("aadhaarName"),
        aadhaar_dob,
        data.get("digilockerRef") or _now_ts(),
        "VERIFIED",
        now,
    )
    await db.execute('UPDATE "User" SET role=$1 WHERE id=$2', "WORKER", user_id)

    profile = await db.fetchrow(
        """
        SELECT wp.*, u.id AS "user_id", u.name AS "user_name", u.phone AS "user_phone"
        FROM "WorkerProfile" wp
        JOIN "User" u ON u.id = wp."userId"
        WHERE wp.id=$1
        """,
        profile_id,
    )
    return _format_worker(profile)


async def get_worker_profile_by_user(user_id: str) -> dict:
    """Resolve the WorkerProfile for a user; 404 if the worker hasn't onboarded."""
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE "userId"=$1', user_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    return dict(profile)


async def update_location(worker_id: str, lat: float, lng: float) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    await db.execute(
        'UPDATE "WorkerProfile" SET latitude=$1, longitude=$2, "updatedAt"=$3 WHERE id=$4',
        lat, lng, now_utc(), worker_id,
    )
    updated = await db.fetchrow(
        """
        SELECT wp.*, u.id AS "user_id", u.name AS "user_name"
        FROM "WorkerProfile" wp JOIN "User" u ON u.id=wp."userId" WHERE wp.id=$1
        """,
        worker_id,
    )
    return _format_worker(updated)


async def set_availability(worker_id: str, is_available: bool, is_on_duty: bool) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    if is_on_duty and profile["status"] != "VERIFIED":
        raise AppError("Worker must be approved before going on duty", 403)
    if is_on_duty and not is_available:
        raise AppError("Cannot be on duty if not available", 400)
    await db.execute(
        'UPDATE "WorkerProfile" SET "isAvailable"=$1, "isOnDuty"=$2, "updatedAt"=$3 WHERE id=$4',
        is_available, is_on_duty, now_utc(), worker_id,
    )
    updated = await db.fetchrow(
        """
        SELECT wp.*, u.id AS "user_id", u.name AS "user_name"
        FROM "WorkerProfile" wp JOIN "User" u ON u.id=wp."userId" WHERE wp.id=$1
        """,
        worker_id,
    )
    return _format_worker(updated)


async def get_worker_profile(worker_id: str) -> dict:
    profile = await db.fetchrow(
        """
        SELECT wp.*, u.id AS "user_id", u.name AS "user_name", u.phone AS "user_phone",
               u."avatarUrl" AS "user_avatarUrl", c.id AS "coop_id", c.name AS "coop_name"
        FROM "WorkerProfile" wp
        JOIN "User" u ON u.id=wp."userId"
        LEFT JOIN "CoOp" c ON c.id=wp."coopId"
        WHERE wp.id=$1
        """,
        worker_id,
    )
    if profile is None:
        raise AppError("Worker profile not found", 404)

    reviews = await db.fetch(
        """
        SELECT r.*, a.name AS "author_name", b.id AS "booking_id", b."bookingRef" AS "booking_bookingRef"
        FROM "Review" r
        JOIN "User" a ON a.id=r."authorId"
        JOIN "Booking" b ON b.id=r."bookingId"
        WHERE r."workerId"=$1
        ORDER BY r."createdAt" DESC LIMIT 10
        """,
        worker_id,
    )

    result = _format_worker(profile)
    result["user"] = {
        "id": profile["user_id"], "name": profile["user_name"],
        "phone": profile["user_phone"], "avatarUrl": profile["user_avatarUrl"],
    }
    result["coop"] = (
        {"id": profile["coop_id"], "name": profile["coop_name"]}
        if profile["coop_id"] else None
    )
    result["reviewsReceived"] = [{
        **dict(r),
        "author": {"id": r["author_name"] and dict(r)["authorId"], "name": r["author_name"]},
        "booking": {"id": r["booking_id"], "bookingRef": r["booking_bookingRef"]},
    } for r in reviews]
    return deep_serialize(result)


async def get_worker_earnings(worker_id: str) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)

    completed = await db.fetch(
        """
        SELECT b.id, b."bookingRef", b."finalPrice", b."workerPayout", b."completedAt",
               s.name AS "service_name", s."categoryName" AS "service_categoryName"
        FROM "Booking" b JOIN "Service" s ON s.id=b."serviceId"
        WHERE b."workerId"=$1 AND b.status='COMPLETED'
        ORDER BY b."completedAt" DESC
        """,
        worker_id,
    )
    transactions = await db.fetch(
        'SELECT * FROM "WalletTransaction" WHERE "workerId"=$1 ORDER BY "createdAt" DESC LIMIT 50',
        worker_id,
    )

    now = now_utc()
    month_start = now.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    monthly = sum(
        num(b["workerPayout"]) or 0
        for b in completed if b["completedAt"] and b["completedAt"] >= month_start
    )

    recent_bookings = []
    for b in completed[:10]:
        recent_bookings.append({
            "id": b["id"], "bookingRef": b["bookingRef"],
            "finalPrice": num(b["finalPrice"]) or 0, "workerPayout": num(b["workerPayout"]) or 0,
            "completedAt": b["completedAt"], "service": {"name": b["service_name"], "categoryName": b["service_categoryName"]},
        })
    recent_transactions = [
        {"id": t["id"], "type": t["type"], "amount": num(t["amount"]),
         "balanceAfter": num(t["balanceAfter"]), "description": t["description"],
         "createdAt": t["createdAt"]}
        for t in transactions[:20]
    ]
    return deep_serialize({
        "totalEarnings": num(profile["totalEarnings"]), "walletBalance": num(profile["walletBalance"]),
        "totalJobs": profile["totalJobs"], "avgRating": num(profile["avgRating"]),
        "monthlyEarnings": round(monthly * 100) / 100,
        "recentBookings": recent_bookings,
        "recentTransactions": recent_transactions,
    })


async def verify_worker(worker_id: str, aadhaar_number: str) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    if not re.fullmatch(r"\d{12}", aadhaar_number):
        raise AppError("Aadhaar verification failed: invalid number", 400)
    await db.execute(
        'UPDATE "WorkerProfile" SET status=$1, "kycStatus"=$2, "aadhaarVerified"=true, "digilockerRef"=$3, "updatedAt"=$4 WHERE id=$5',
        "VERIFIED", "VERIFIED", _now_ts(), now_utc(), worker_id,
    )
    updated = await db.fetchrow(
        'SELECT wp.*, u.id AS "user_id", u.name AS "user_name" FROM "WorkerProfile" wp JOIN "User" u ON u.id=wp."userId" WHERE wp.id=$1',
        worker_id,
    )
    return _format_worker(updated)


async def approve_worker(worker_id: str, coop_id: str, note: str | None = None) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    if profile["coopId"] and profile["coopId"] != coop_id:
        raise AppError("Worker does not belong to this co-op", 403)
    if profile["status"] in ("SUSPENDED", "DEACTIVATED"):
        raise AppError("Cannot approve a suspended/deactivated worker", 400)
    await db.execute(
        'UPDATE "WorkerProfile" SET status=$1, "kycStatus"=$2, "aadhaarVerified"=true, "updatedAt"=$3 WHERE id=$4',
        "VERIFIED", "VERIFIED", now_utc(), worker_id,
    )
    updated = await db.fetchrow(
        'SELECT wp.*, u.id AS "user_id", u.name AS "user_name" FROM "WorkerProfile" wp JOIN "User" u ON u.id=wp."userId" WHERE wp.id=$1',
        worker_id,
    )
    return _format_worker(updated)


async def reject_worker(worker_id: str, coop_id: str, reason: str) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    if profile["coopId"] and profile["coopId"] != coop_id:
        raise AppError("Worker does not belong to this co-op", 403)
    await db.execute(
        'UPDATE "WorkerProfile" SET status=$1, "kycStatus"=$2, "updatedAt"=$3 WHERE id=$4',
        "SUSPENDED", "REJECTED", now_utc(), worker_id,
    )
    updated = await db.fetchrow(
        'SELECT wp.*, u.id AS "user_id", u.name AS "user_name" FROM "WorkerProfile" wp JOIN "User" u ON u.id=wp."userId" WHERE wp.id=$1',
        worker_id,
    )
    return _format_worker(updated)


async def search_workers(lat: float, lng: float, radius_km: float,
                         skill_tags=None, coop_id: str | None = None) -> list[dict]:
    rows = await db.fetch(
        """
        SELECT wp.*, u.name AS "user_name", c.name AS "coop_name"
        FROM "WorkerProfile" wp
        JOIN "User" u ON u.id=wp."userId"
        LEFT JOIN "CoOp" c ON c.id=wp."coopId"
        WHERE wp.status='VERIFIED' AND wp.latitude IS NOT NULL AND wp.longitude IS NOT NULL
          AND u."isActive"=true
          AND ($1::text IS NULL OR wp."coopId"=$1)
        LIMIT 100
        """,
        coop_id,
    )
    result = []
    for w in rows:
        d = haversine_km(lat, lng, num(w["latitude"]), num(w["longitude"]))
        item = {
            "workerId": w["id"], "workerName": w["user_name"], "coopName": w["coop_name"] or None,
            "skillTags": w["skillTags"], "avgRating": num(w["avgRating"]), "totalJobs": w["totalJobs"],
            "bio": w["bio"], "experienceYears": w["experienceYears"],
            "isAvailable": w["isAvailable"], "isOnDuty": w["isOnDuty"],
            "distanceKm": round(d * 100) / 100,
        }
        if not matches_skills(w["skillTags"], skill_tags):
            continue
        if item["distanceKm"] > radius_km:
            continue
        item["etaMinutes"] = max(10, round(item["distanceKm"] * 5))
        item["matchScore"] = worker_match_score(item["avgRating"], item["totalJobs"], item["distanceKm"], radius_km)
        result.append(item)
    result.sort(key=lambda x: (-x["matchScore"], x["distanceKm"]))
    return deep_serialize(result[:100])


def _format_worker(w: dict) -> dict:
    out = dict(w)
    total_earnings = num(w.get("totalEarnings"))
    wallet = num(w.get("walletBalance"))
    if total_earnings is not None:
        out["totalEarnings"] = total_earnings
    if wallet is not None:
        out["walletBalance"] = wallet
    for key in ("user_id", "user_name", "user_phone", "user_avatarUrl", "coop_id", "coop_name"):
        out.pop(key, None)
    return deep_serialize(out)


def num_or(v, d=0):
    n = num(v)
    return d if n is None else n
